platforms/jvm/plugins-java-base/src/main/java/org/gradle/api/plugins/internal/JvmPluginsHelper.java

```diff
@@ -177,21 +179,44 @@ public static Configuration createDocumentationVariantWithArtifact(
      * @param compatibilityComputer A function to compute the compatibility version to use as the convention
      *      given the raw version values and the current version property values set on the extension
      */
-    public static void configureCompileDefaults(AbstractCompile compile, DefaultJavaPluginExtension javaExtension, BiFunction<JavaVersion, Supplier<JavaVersion>, JavaVersion> compatibilityComputer) {
-        ConventionMapping conventionMapping = compile.getConventionMapping();
-        conventionMapping.map("sourceCompatibility", () -> computeSourceCompatibilityConvention(javaExtension, compatibilityComputer).toString());
-        conventionMapping.map("targetCompatibility", () -> computeTargetCompatibilityConvention(javaExtension, compile, compatibilityComputer).toString());
+    public static void configureCompileDefaults(AbstractCompile compile, DefaultJavaPluginExtension javaExtension, BiFunction<Provider<JavaVersion>, Provider<JavaVersion>, Provider<JavaVersion>> compatibilityComputer) {
+        compile.getSourceCompatibility().convention(
+            compatibilityComputer.apply(
+                javaExtension.getSourceCompatibility(),
+                javaExtension.getEffectiveSourceCompatibility()
+            ).map(JavaVersion::toString)
+        );
+        compile.getTargetCompatibility().convention(
+            compatibilityComputer.apply(
+                javaExtension.getTargetCompatibility()
+                    .orElse(javaExtension.getSourceCompatibility())
+                    .orElse(getRawSourceCompatibilityValue(compile).map(JavaVersion::toVersion)),
+                javaExtension.getEffectiveTargetCompatibility()
+            ).map(JavaVersion::toString)
+        );
     }
 
-    private static JavaVersion computeSourceCompatibilityConvention(DefaultJavaPluginExtension javaExtension, BiFunction<JavaVersion, Supplier<JavaVersion>, JavaVersion> compatibilityComputer) {
-        return compatibilityComputer.apply(javaExtension.getRawSourceCompatibility(), javaExtension::getSourceCompatibility);
-    }
+    // we need to know value WITHOUT convention before we set the convention
```

Comment:
See 
https://github.com/gradle/gradle/issues/31886
for a way to improve this

Reply:
I think we should do what we can to avoid this -- or any dependency on needing the "raw" value of the source/target compatibility. 

In this case, I think this logic was working around the fact that `getTargetCompatibility` previously returned the _computed_ (effective) target compatibility. 

Now, the `getTargetCompatibility` always provide the raw user-provided value, so I don't think this logic is necessary at all

Reply:
The issue here is that the old logic relies on the configured value (with conventions to override). 
Once we set the convention for the _task_ source compatibility, we can't use the configured value anymore. This means that the _task_target compatibility will be resolved in a different order.
We can introduce the `effective` counterpart for the task, too, but it defeats the purpose IMHO. The alternative to that is to redesign the conventions, but this solves a different problem.